cmd/ocsp-responder/main.go

```diff
@@ -32,27 +33,101 @@ import (
+	if len(f.serialPrefixes) > 0 {
+		match := false
+		for _, prefix := range f.serialPrefixes {
+			if match = strings.HasPrefix(serialString, prefix); match {
+				break
+			}
+		}
+		if !match {
+			return fmt.Errorf("Request serial has wrong prefix: %w", bocsp.ErrNotFound)
+		}
+	}
 
-  CREATE TABLE ocsp_responses (serialNumber TEXT, response BLOB);
+	return nil
+}
 
-The serialNumber field may have any type to which Go will match a string,
-so you can be more efficient than TEXT if you like.  We use it to store the
-serial number in base64.  You probably want to have an index on the
-serialNumber field, since we will always query on it.
+func newFilter(issuerCerts []string, serialPrefixes []string) (*ocspFilter, error) {
+	if len(issuerCerts) < 1 {
+		return nil, errors.New("Filter must include at least 1 issuer cert")
+	}
+	var issuerKeyHashes [][]byte
+	for _, issuerCert := range issuerCerts {
+		// Load the certificate from the file path.
+		caCertDER, err := cmd.LoadCert(issuerCert)
+		if err != nil {
+			return nil, fmt.Errorf("Could not read issuer cert %s: %w", issuerCert, err)
+		}
+		caCert, err := x509.ParseCertificate(caCertDER)
+		if err != nil {
+			return nil, fmt.Errorf("Could not parse issuer cert %s: %w", issuerCert, err)
+		}
+		// The issuerKeyHash in OCSP requests is constructed over the DER
+		// encoding of the public key per RFC 6960 (defined in RFC 4055 for
+		// RSA and RFC  5480 for ECDSA). We can't use MarshalPKIXPublicKey
+		// for this since it encodes keys using the SPKI structure itself,
+		// and we just want the contents of the subjectPublicKey for the
+		// hash, so we need to extract it ourselves.
+		var spki struct {
+			Algo      pkix.AlgorithmIdentifier
+			BitString asn1.BitString
+		}
+		if _, err := asn1.Unmarshal(caCert.RawSubjectPublicKeyInfo, &spki); err != nil {
+			return nil, err
+		}
+		keyHash := sha1.Sum(spki.BitString.Bytes)
+		issuerKeyHashes = append(issuerKeyHashes, keyHash[:])
+	}
+	return &ocspFilter{crypto.SHA1, issuerKeyHashes, serialPrefixes}, nil
+}
 
-*/
-type DBSource struct {
-	dbMap             dbSelector
-	caKeyHash         []byte
-	reqSerialPrefixes []string
-	timeout           time.Duration
-	log               blog.Logger
+// dbSource represents a database containing pre-generated OCSP responses keyed
+// by serial number. It also allows for filtering requests by their issuer key
+// hash and serial number, to prevent unnecessary lookups for rows that we know
+// will not exist in the database.
+//
+// We assume that OCSP responses are stored in a very simple database table,
+// with two columns: serialNumber and response
+//
+// 	CREATE TABLE ocsp_responses (serialNumber TEXT, response BLOB);
+//
+// The serialNumber field may have any type to which Go will match a string,
+// so you can be more efficient than TEXT if you like.  We use it to store the
+// serial number in base64.  You probably want to have an index on the
```

Comment:
This is also outdated; serial number is stored as hex. And I think we can change "probably want to have an index" to "must have an index." 😂

Reply:
Done.